src/main/java/muramasa/antimatter/capability/CapabilityHandler.java

```diff
@@ -0,0 +1,120 @@
+package muramasa.antimatter.capability;
+
+import muramasa.antimatter.Antimatter;
+import muramasa.antimatter.network.packets.CapabilityPacket;
+import muramasa.antimatter.tile.TileEntityBase;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraftforge.api.distmarker.Dist;
+
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+public class CapabilityHandler<T extends TileEntityBase, H extends ICapabilityHandler> {
+
+    protected T tile;
+    protected H handler;
+    protected CompoundNBT tag;
+    protected BiFunction<T, CompoundNBT, H> capability = (tile, tag) -> null;
+    protected Dist side; // when null, works on both sides
```

Comment:
I don't quite like this solution for determine existence of capabilities. Since `Dist::DEDICATED_SERVER` doesn't mean server-side, but dedicated server side, this means it should not exist in integrated server,

Also I have been highly advised to only send packets to client when they're either

1. Watching for the capability
2. To notify any client rendering changes

Packets should be sent sparingly.

Reply:
I replace Dist by a new enum class